### PR TITLE
RFC: Memoize style objects to prevent re-renders of memoized components

### PR DESCRIPTION
**Pull request by levibuzolic**
Awesome work Jacek on a great library, I love the simple approach and it fits well with patterns I've used across a few apps already and I'm looking forward to seeing how it evolves.

I'd like to request a feature/open a discussion about how to better handle styles in a way that doesn't break memoization of components (ie. `React.memo()`)

# Problem

While integrating `react-native-unistyles` into a project we've noticed that styles break memoization (`React.memo`) when _any_ dynamic features are used (ie. anything requiring `useStyles()`). This is because the `useStyles` hook returns a new style object on every render, which causes the memoized component to re-render every time.

It does however work as expected when using purely static styles because no internal manipulation is done to the passed-in style object.

This can be demonstrated by this [Expo Snack](https://snack.expo.dev/@levibuzolic/unistyles-static). Where every press of the `Re-render` button triggers both dynamic style components to re-render.

image

# User-land solution

To work around this issue we've been creating a separate static style object alongside our dynamic styles, or avoided passing styles entirely. (As demonstrated in the above Expo Snack)

```tsx
// Dynamic styles, all unsafe to pass through to a memoized component
const stylesheet = createStyleSheet(({colors}) => ({
  item: {
    backgroundColor: colors.fg.primary,
  }
  text: {
    fontSize: 12,
  },
}));

// Static styles, safe to use in memoized components
const staticStyles = createStyleSheet({
  other: {
    borderRadius: 4,
  }
});
```

This means we haven't been able to get the full benefits of Unistyles in the few heavily optimised areas of our codebase.

# Proposed solution (contents of this PR)

A few targetted uses of `useMemo` throughout the `useStyles` hook means that we can make styles memoizable by default, and only recompute them when either the screen size or the theme changes. 

This has a trade-off of potentially increased memory usage, though at the same time would significantly reduce the amount of object transforms (nested `parseStyle` calls) on every render.

Kapture 2023-10-12 at 00 23 04

### CODE REVIEW

File: src/hooks/useDimensions.ts [hunk 1]
```diff
@@ -1,11 +1,4 @@
...
 
-export const useDimensions = (): ScreenSize => {
-    const { width, height } = useWindowDimensions()
-
-    return {
-        width,
-        height
-    }
-}
+export const useDimensions = (): ScreenSize => useWindowDimensions()
```
Comment (levibuzolic) [inline src/hooks/useDimensions.ts:4]: Unwrapping this object and creating a new object breaks memoization. I was going to reach for another `useMemo` here, but the object is already stable direct from React Native, so we might as well consume it directly.

**Comment by jpudysz** [comment]
Wow @levibuzolic , thank you a lot for this PR!

I need some time to test it, and I will go back to you (most likely tomorrow).

File: src/createUnistyles.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { useContext } from 'react'
+import { useContext, useMemo, useEffect, useRef } from 'react'
```
Comment (jpudysz) [inline src/createUnistyles.ts:1]: remove unused `useEffect` and `useRef`

File: src/createUnistyles.ts [hunk 3]
```diff
@@ -57,7 +58,7 @@ export const createUnistyles = <B extends Breakpoints, T = {}>(breakpoints: B) =
                         ...acc,
                         [key]: parseStyle<ST, B>(style, breakpoint, screenSize, sortedBreakpointEntries)
                     })
-                }, {} as ST)
+                }, {} as ST), [breakpoint, screenSize, sortedBreakpointEntries])
```
Comment (jpudysz) [inline src/createUnistyles.ts:61]: you don't need to pass `sortedBreakpointEntries` to the deps. It's computed only once, before uses `useStyles`. Value will never change.

**Review by jpudysz — CHANGES_REQUESTED** [review]
After the initial tests, I must say that this is an excellent PR. Thank you for delving into it. It will greatly improve the library!

File: src/createUnistyles.ts (hunk 3 above)
Comment (efstathiosntonas) [reply src/createUnistyles.ts:61]: `theme` needs to be passed to the deps array too otherwise theme won't be changed after user changes it

Comment (levibuzolic) [reply src/createUnistyles.ts:61]: Ah, of course. It looks like `parsedStyles` is actually what should be passed in here, that's the actual dependency which relies on `theme` and `stylesheet`.

**Comment by levibuzolic** [comment]
Thanks for catching those issues @jpudysz, should be sorted now. :+1:

File: src/createUnistyles.ts (hunk 3 above)
Comment (efstathiosntonas) [reply src/createUnistyles.ts]: yeah you're 💯 right! Thanks @levibuzolic

**Review by jpudysz — APPROVED** [review]

**Comment by jpudysz** [comment]
I will merge it and fix it in a separate PR as your timezone is 9h ahead 😅 .
Thank you @levibuzolic for the PR

**Pull request merged**